pkg/controller/velerobackup/hiveobject_test.go

```diff
@@ -0,0 +1,185 @@
+			expectedHiveObject: nil,
+			expectedError:      fmt.Errorf("Unknown Type: *v1.Pod"),
+		},
+		{
+			name:               "Invalid nil object",
+			object:             nil,
+			expectedHiveObject: nil,
+			expectedError:      fmt.Errorf("Can't convert nil to a HiveObject"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+
+			// Act
+			actualHiveObject, actualError := newHiveObject(test.object, test.logger)
+
+			// Assert
+			assert.Equal(t, test.expectedHiveObject, actualHiveObject)
+			assert.Equal(t, test.expectedError, actualError)
+		})
+	}
+}
+
+func TestHasChanged(t *testing.T) {
+	tests := []struct {
+		name            string
+		hiveObject      *hiveObject
+		expectedChanged bool
+		checksumFunc    controllerutils.ChecksumOfObjectsFunc
+	}{
+		{
+			name:            "Unchanged ClusterDeployment",
+			hiveObject:      ro2ho(testclusterdeployment.Build(unchangedClusterDeploymentBase())),
+			expectedChanged: false,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Changed ClusterDeployment",
+			hiveObject:      ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			expectedChanged: true,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Unchanged SyncSet",
+			hiveObject:      ro2ho(testsyncset.Build(unchangedSyncSetBase())),
+			expectedChanged: false,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Changed SyncSet",
+			hiveObject:      ro2ho(testsyncset.Build(changedSyncSetBase())),
+			expectedChanged: true,
+			checksumFunc:    controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:            "Error from Checksum func",
+			hiveObject:      ro2ho(testsyncset.Build(unchangedSyncSetBase())),
+			expectedChanged: false,
+			checksumFunc:    getHashOfObjectsFuncReturnsError,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+
+			// Act
+			actualChanged := test.hiveObject.hasChanged()
+
+			// Assert
+			assert.Equal(t, test.expectedChanged, actualChanged)
+		})
+	}
+}
+
+func TestCalculateChecksum(t *testing.T) {
+	tests := []struct {
+		name             string
+		checksumObject   *hivev1.ClusterDeployment
+		expectedChecksum string
+		checksumFunc     controllerutils.ChecksumOfObjectsFunc
+	}{
+		{
+			name:             "Valid ClusterDeployment Checksum",
+			checksumObject:   testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+			expectedChecksum: testclusterdeployment.Build(unchangedClusterDeploymentBase()).Annotations[controllerutils.LastBackupAnnotation],
+			checksumFunc:     controllerutils.GetChecksumOfObjects,
+		},
+		{
+			name:             "Error from Checksum func",
+			checksumObject:   testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+			expectedChecksum: errChecksum,
+			checksumFunc:     getHashOfObjectsFuncReturnsError,
```

Comment:
Rather than having to mock out the checksum function, rework this test so that the test cases specifies the meta and spec as inputs and use a spec value that cannot be marshalled. Then you can get rid of the `checksumFunc` parameter entirely.

For example,
```
tests := []struct{
  name string
  meta *metav1.Object
  spec interface{}
  expectedChecksum string
}{
  {
    name: "Error from Checksum func",
    meta: &metav1.ObjectMeta{},
    spec: math.Inf(1),
    expectedChceksum: errChecksum,
  },
}
```